### DISCUSSION TITLE
cyclicAMI | FOAM FATAL ERROR: Unable to find initial target face

### DISCUSSION THREAD

**Discussion by raisePiginYourhome** [mdolab/dafoam#466] [General]
i use the dasimplefoam to optimise a centrifugal impeller with cyclicAMI peridic boundary condition, when a loop starts to run, the FOAM FATAL ERROR shown in the console, how to fix it? Thanks alot for your assistance.

**Comment by uttamcadambi07** [comment]
I think it is mentioned in the FAQs that cyclicAMI is not supported. Here
```
Does DAFoam support all the OpenFOAM’s configurations and boundary conditions?

No. These OpenFOAM features are NOT supported in DAFoam. Note: the configurations and boundary conditions used in [DAFoam tutorials](https://github.com/dafoam/tutorials) are tested and working. Use caution if you want to add a new configuration that has not been used the tutorials.

    Unsteady solvers
    AMI boundary condition
    Non-stationary boundary conditions, such as totalPressure, totalTemperature
    fvOptions and MRF are implemented for only some of primal solvers
    empty boundary condition
    Limited schemes such as Gauss linear limited corrected 0.33 for laplacianSchemes and limited corrected 0.33 for snGradSchemes in system/fvSchemes may cause inaccurate adjoint gradients. Don’t use them
```
https://dafoam.github.io/mydoc_get_started_faq.html

**Reply by raisePiginYourhome** [reply]
got it,thanks

**Reply by raisePiginYourhome** [reply]
Is there any softwares could draw the cyclic mesh, i have tried fluent,NETGEN, but these can'r pass the meshcheck.

**Reply by uttamcadambi07** [reply]
what do you mean by cyclic mesh? Do you mean by cyclic boundary condition? There are cyclic boundary patches depending on the geometry but they are not related to the mesh. When it does not pass the mesh check, what error do you encounter?

**Reply by raisePiginYourhome** [reply]
- --> FOAM FATAL ERROR: (openfoam-2212 patch=230110)
face 0 area does not match neighbour by 25.2168% -- possible face ordering problem.
patch:per1 my area:7.76066e-06 neighbour area:1e-05 matching tolerance:1e-07
Mesh face:46429 fc:(-0.0200771 0.0229884 -0.0351103)

**Reply by raisePiginYourhome** [reply]
when i incressing the matching tolerance, it dosesn't help.

**Reply by uttamcadambi07** [reply]
https://www.cfd-online.com/Forums/openfoam-meshing/133730-problem-cyclic-boundaries-openfoam-2-3-mesh-import-icem.html
Check this thread. This seems like an OpenFOAM problem so I doubt you would get much support here.

**Reply by raisePiginYourhome** [reply]
So I think it's the mesh quality that can't satisfy the cyclic boundary condition,is there any way to deal this problem.

**Reply by raisePiginYourhome** [reply]
ok, i'll check this, thanks

**Reply by raisePiginYourhome** [reply]
This thread gave the solution that change type cyclic to cyclicAMI, but DAFoam seems not surpoort the boundary type cyclicAMI, so if i can improve the mesh quality or there is another way to address the issue, could you provide more solutions about this error. 
Thank you really and look forward to your response.

**Comment by friedenhe** [comment]
@raisePiginYourhome The error you saw is likely related to your cyclicAMI setup in OpenFOAM instead of DAFoam. You need to run a native OpenFOAM solver (e.g., simpleFoam) and see if you see the same error. If yes, the problem is not caused by DAFoam. 

The problem that cyclicAMI is not supported is that the cyclicAMI BC is not AD properly in parallel, which will degrade the adjoint accuracy. But it will not give you any errors. So, you can run an optimization with the cyclicAMI BC in serial. If you need to run it in parallel, a workaround is to use the ["singleProcessorFaceSets"](https://github.com/mdolab/dafoam/blob/a00afb1261888acd67d4945a8affc84b70e15fd2/dafoam/pyDAFoam.py#L603C14-L603C37) to force one processor to own two cyclicAMI patches. Note that the "singleProcessorFaceSets" trick should work in theory, but we have not tested it. If it works for you, please let us know.

BTW: ICEM-CFD and Pointwise can generate nice cyclic meshes for OpenFOAM.

**Reply by raisePiginYourhome** [reply]
Thank you, PingHe.

**Reply by raisePiginYourhome** [reply]
I have tried to use ICEM to make a mesh for OpenFOAM

![mesh_screenCut](https://github.com/mdolab/dafoam/assets/126868299/d61c3268-c857-4b23-8afb-05bade1d583e)

When i set the cyclicAMI boundary condition for periodic SimpleFOAM work well, and the simulation results are also very accurate.
But this mesh seems can't pass the meshcheck which contains zero or negative cell volume,non-orthogonality and face pyramids. So i think is there any way to pass the meshcheck,.

Looking forward for your advices. Thanks alot.

![Failed_meshchecks pyDAFoamErr](https://github.com/mdolab/dafoam/assets/126868299/da1a9cbe-cf9e-4c47-9c01-f5c9c1650639)

**Reply by raisePiginYourhome** [reply]
this is the openfoam2212 meshcheck, and it passes the mesh check.
![openfoamMeshCheck](https://github.com/mdolab/dafoam/assets/126868299/88c5c1c0-faec-4011-ad97-5ebfe2c43d14)

**Reply by friedenhe** [reply]
I suspect your FFD does not fully contain the design surface. Can you post a picture of your design surface overlapped with the FFD points in Paraview?

**Reply by raisePiginYourhome** [reply]
![FFD_points](https://github.com/mdolab/dafoam/assets/126868299/8bb0509c-dd21-4b71-81c5-1f22e3cf4b1d)
Here are the FFD points, by the way are you Chinese, Ping He?

**Reply by raisePiginYourhome** [reply]
It seems DASimpleFoam dont't support totalpressureRatio type for inlet and outlet, so i choosed the type totalPressure to set physical constrains, is that workable?

**Reply by raisePiginYourhome** [reply]
and the default pressure unit is kpa or pa?

**Reply by friedenhe** [reply]
The pressure unit is pa. Are you sure your FFD fully contains the design surface? From the picture you sent, it is hard to tell. Also, you can check which part of the mesh has poor quality. check the DAFoam website FAQ.

**Reply by raisePiginYourhome** [reply]
OK, I'll check this
